lsp/nls/src/command.rs

```diff
@@ -0,0 +1,39 @@
+use lsp_server::{RequestId, Response, ResponseError};
+use lsp_types::{ExecuteCommandParams, TextDocumentIdentifier, Url};
+use nickel_lang_core::{
+    error::IntoDiagnostics,
+    eval::{cache::CacheImpl, VirtualMachine},
+};
+
+use crate::{cache::CacheExt, error::Error, server::Server};
+
+pub fn handle_command(
+    params: ExecuteCommandParams,
+    req: RequestId,
+    server: &mut Server,
+) -> Result<(), ResponseError> {
+    match params.command.as_str() {
+        "eval" => {
+            server.reply(Response::new_ok(req, None::<()>));
+
+            let doc: TextDocumentIdentifier =
+                serde_json::from_value(params.arguments[0].clone()).unwrap();
+            eval(server, &doc.uri)?;
+            Ok(())
+        }
+        _ => Err(Error::CommandNotFound(params.command).into()),
+    }
+}
+
+fn eval(server: &mut Server, uri: &Url) -> Result<(), Error> {
+    if let Some(file_id) = server.cache.file_id(uri)? {
+        // TODO: avoid cloning the cache. Maybe we can have a VM with a &mut ImportResolver?
```

Comment:
Hmm, this would be annoying because we'd have to add a lifetime parameter to `VirtualMachine`. I wonder if you can't just implement `ImportResolver` for `&'a mut R` when `R: ImportResolver`, thus passing a mutable reference instead? Or just wrap it as a dedicated type `ImportResolverRef<'a, R: ImportResolver>(&'a mut ImportResolver)` and implement the trait. After all `VirtualMachine` is generic over the resolver.

Reply:
Oh, that's a good point. I'll give it a quick try, and if it works I'll add it to this PR

Reply:
Implementing `ImportResolver` for `&'a mut R` worked nicely, but it didn't solve the issue because `prepare_eval` is only defined when the import resolver parameter is `ImportCache`. I'll just leave the TODO for now...